### PR TITLE
add a new query laning metrics to visualize lane assignment

### PR DESCRIPTION
**Pull request by isandeep41**
This PR aims to introduce a new metrics `query/priority` which will help to visualize the assigned lane if the [Laning Strategy][1] is enabled.

image-20220103155445805
The Grafana Dashboard shows the distribution of `high/low` lanes on a Druid cluster running Druid version 0.22.1

[1]: https://druid.apache.org/docs/latest/configuration/index.html#laning-strategies "Laning strategies"

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/QuerySchedulerProvider.java [hunk 1]
```diff
@@ -23,27 +23,28 @@
...
 import com.google.inject.Provider;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
 import org.apache.druid.server.initialization.ServerConfig;
 
 
 public class QuerySchedulerProvider extends QuerySchedulerConfig implements Provider<QueryScheduler>
 {
   private final ServerConfig serverConfig;
-
+  private ServiceEmitter emitter;
```
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/QuerySchedulerProvider.java:33]:
```suggestion
  private final ServiceEmitter emitter;
```

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java [hunk 2]
```diff
@@ -137,6 +170,12 @@ public void registerQueryFuture(Query<?> query, ListenableFuture<?> future)
     Optional<Integer> priority = prioritizationStrategy.computePriority(queryPlus, segments);
     query = priority.map(query::withPriority).orElse(query);
     Optional<String> lane = laningStrategy.computeLane(queryPlus.withQuery(query), segments);
+    log.info("[%s] lane assigned to [%s] query with [%,d] priority", lane.orElse("default"), query.getType(), priority.orElse(new Integer(0)));
+    final ServiceMetricEvent.Builder builderUsr = ServiceMetricEvent.builder().setFeed("metrics")
+                                                                    .setDimension("lane", lane.orElse("default"))
+                                                                    .setDimension("dataSource", query.getDataSource().getTableNames())
+                                                                    .setDimension("type", query.getType());
+    emitter.emit(builderUsr.build("query/priority", priority.orElse(new Integer(0))));
```
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/QueryScheduler.java:178]:
```suggestion
    emitter.emit(builderUsr.build("query/priority", priority.orElse(0)));
```

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java [hunk 3]
```diff
@@ -86,7 +91,34 @@
...
+      this.totalCapacity = totalNumThreads;
+    } else {
+      limitTotal = false;
+      this.totalCapacity = serverConfig.getNumThreads();
+    }
+    this.laneRegistry = BulkheadRegistry.of(getLaneConfigs(limitTotal));
+    this.emitter = emitter;
+  }
+
+  @VisibleForTesting
```
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/QueryScheduler.java:121]: Why is this annotation needed? I see the ctor is declared as public

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java [hunk 4]
```diff
@@ -108,6 +140,7 @@ public QueryScheduler(
       this.totalCapacity = serverConfig.getNumThreads();
     }
     this.laneRegistry = BulkheadRegistry.of(getLaneConfigs(limitTotal));
+    this.emitter = new ServiceEmitter("test", "localhost", new NoopEmitter());
```
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/QueryScheduler.java:143]: So this ctor is for the test case only? I think we'd better not instantiate such object here but pass an object from the test cases.

Comment (isandeep41) [reply server/src/main/java/org/apache/druid/server/QueryScheduler.java]: agree, have removed the ctor and the object is now directly initiated from the test classes

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java (hunk 3 above)
Comment (isandeep41) [reply server/src/main/java/org/apache/druid/server/QueryScheduler.java]: the ctor has been removed

File: docs/operations/metrics.md [hunk 5]
```diff
@@ -58,6 +58,7 @@ Available Metrics
 |`query/interrupted/count`|number of queries interrupted due to cancellation.|This metric is only available if the QueryCountStatsMonitor module is included.||
 |`query/timeout/count`|number of timed out queries.|This metric is only available if the QueryCountStatsMonitor module is included.||
 |`query/segments/count`|This metric is not enabled by default. See the `QueryMetrics` Interface for reference regarding enabling this metric. Number of segments that will be touched by the query. In the broker, it makes a plan to distribute the query to realtime tasks and historicals based on a snapshot of segment distribution state. If there are some segments moved after this snapshot is created, certain historicals and realtime tasks can report those segments as missing to the broker. The broker will re-send the query to the new servers that serve those segments after move. In this case, those segments can be counted more than once in this metric.|Varies.|
+|`query/priority`| Assigned lane and priority, only if Laning strategy is enabled. Refer to [Laning strategies](../configuration/index.md#laning-strategies)| lane, dataSource, type|0|
```
Comment (asdf2014) [inline docs/operations/metrics.md:61]:
```suggestion
|`query/priority`|Assigned lane and priority, only if Laning strategy is enabled. Refer to [Laning strategies](../configuration/index.md#laning-strategies)|lane, dataSource, type|0|
```

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java [hunk 6]
```diff
@@ -58,6 +62,7 @@
  */
 public class QueryScheduler implements QueryWatcher
 {
+  private static final Logger log = new Logger(QueryScheduler.class);
```
Comment (asdf2014) [inline server/src/main/java/org/apache/druid/server/QueryScheduler.java:65]:
```suggestion
  private static final Logger LOGGER = new Logger(QueryScheduler.class);
```

File: server/src/main/java/org/apache/druid/server/QueryScheduler.java [hunk 7]
```diff
@@ -137,6 +159,12 @@ public void registerQueryFuture(Query<?> query, ListenableFuture<?> future)
     Optional<Integer> priority = prioritizationStrategy.computePriority(queryPlus, segments);
     query = priority.map(query::withPriority).orElse(query);
     Optional<String> lane = laningStrategy.computeLane(queryPlus.withQuery(query), segments);
+    log.info("[%s] lane assigned to [%s] query with [%,d] priority", lane.orElse("default"), query.getType(), priority.orElse(Integer.valueOf(0)));
```
Comment (asdf2014) [inline server/src/main/java/org/apache/druid/server/QueryScheduler.java:162]:
```suggestion
    LOGGER.info("[%s] lane assigned to [%s] query with [%,d] priority", lane.orElse("default"), query.getType(), priority.orElse(Integer.valueOf(0)));
```

**Review by asdf2014 — APPROVED** [review]
:+1: Overall LGTM, thanks for your contirubtion, also I leave a few small suggestions

**Review by FrankChen021 — APPROVED** [review]

**Comment by jihoonson** [comment]
@isandeep41 @FrankChen021 @asdf2014, thank you for getting this PR merged. It seems that Travis has started to fail after this PR is merged (https://app.travis-ci.com/github/apache/druid/builds/247395466). Can you please check the Travis failure? This is a blocker issue since the CI is broken.

**Comment by gianm** [comment]
I included a fix in #12307. It was a logical conflict between this patch and #12202, that as far as I can tell only affected the test case QueryableModuleTest.

**Comment by jihoonson** [comment]
Thank you @gianm!

**Comment by FrankChen021** [comment]
Thank you @gianm

**Comment by isandeep41** [comment]
Thanks @gianm

**Pull request merged**